File: 148-sort_list.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Base case: if the list is empty or has only one element, it's already sorted
        if not head or not head.next:
            return head
        
        # Step 1: Find the middle of the list
        mid = self.getMid(head)
        
        # Step 2: Split the list into two halves
        left = head
        right = mid.next
        mid.next = None  # Cut off the first half
        
        # Step 3: Sort each half
        left = self.sortList(left)
        right = self.sortList(right)
        
        # Step 4: Merge the sorted halves
        return self.mergeTwoLists(left, right)
    
    # Helper function to find the middle of the linked list
    def getMid(self, head: ListNode) -> ListNode:
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

    # Helper function to merge two sorted linked lists
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode()
        tail = dummy
        
        while l1 and l2:
            if l1.val < l2.val:
                tail.next = l1
                l1 = l1.next
            else:
                tail.next = l2
                l2 = l2.next
            tail = tail.next
        
        # If one of the lists is not fully merged, append the remaining nodes
        if l1:
            tail.next = l1
        elif l2:
            tail.next = l2
        
        return dummy.next
        

File: 148-sort_list.js
// Definition for singly-linked list.
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val);
    this.next = (next === undefined ? null : next);
}

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function(head) {
    // Base case: if head is null or there is only one node
    if (!head || !head.next) {
        return head;
    }

    // Step 1: Split the list into two halves using slow and fast pointers
    let slow = head;
    let fast = head;
    let prev = null;

    while (fast && fast.next) {
        prev = slow;
        slow = slow.next;
        fast = fast.next.next;
    }

    // Disconnect the first half from the second
    prev.next = null;

    // Step 2: Recursively sort each half
    let left = sortList(head);
    let right = sortList(slow);

    // Step 3: Merge the two sorted halves
    return merge(left, right);
};

// Helper function to merge two sorted lists
function merge(l1, l2) {
    let dummy = new ListNode(0);
    let current = dummy;

    while (l1 && l2) {
        if (l1.val < l2.val) {
            current.next = l1;
            l1 = l1.next;
        } else {
            current.next = l2;
            l2 = l2.next;
        }
        current = current.next;
    }

    // If one list is longer, append the rest of it
    current.next = l1 ? l1 : l2;

    return dummy.next;
}
